# This script contains:
1. Import Libraries
2. Importing Data Sets
3. Aggregating Data with agg()
4. Aggregating Data with Transform()
5. Deriving Columns with loc()
6. Question 5
7. Question 6
8. Question 7
9. Question 9

# 1. Import Libraries

In [2]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# 2.Importing Data Sets

In [5]:
# Defining Path
path= r'/Users/navneetjha/Desktop/Career foundry Docs/exercises/Data Immersion /Achievement 4/INSTACART BASKET ANALYSIS JULY14-2024'

In [7]:
path

'/Users/navneetjha/Desktop/Career foundry Docs/exercises/Data Immersion /Achievement 4/INSTACART BASKET ANALYSIS JULY14-2024'

In [9]:
# Import "orders_products_new_variables_busiesthour_busiestdays" dataframe as "ords_prods_merge" dataframe

ords_prods_merge=pd.read_pickle(os.path.join(path,'02 DATA', 'PREPARED DATA','orders_products_new_variables_busiesthour_busiestdays.pkl'))

In [11]:
# Checking shape of dataset
ords_prods_merge.shape

(32434212, 16)

In [13]:
# Checking top cells
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   3139998      138            28          6                 11   
1   1977647      138            30          6                 17   
2    389851      709             2          0                 21   
3    652770      764             1          3                 13   
4   1813452      764             3          4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  busiest_days  \
0                     3.0                  5          0   both  Regular Days   
1                    20.0                  1          1   both  Regular Days   
2                     6.0                 20          0   both  Busiest Days   
3                     NaN                 10          0   both  Slowest Days   
4                     9.0                 11          1   both  Slowest Days   

  busiest_period_of_day  
0           Most orders  
1        Average Orders  
2        Average Orders  
3           Most orders  
4        Average Orders

# 3. Aggregating Data with agg()

In [16]:
# Grouping Data on "department_id" and aggregating "order_number" to find the mean value
ords_prods_merge.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.213779
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170828
12               15.887622
13               16.583304
14               16.757377
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

In [17]:
#Performing Multiple Aggregations
ords_prods_merge.groupby('department_id').agg({'order_number':['mean','min','max']})

order_number        
                      mean min max
department_id                     
1                15.457687   1  99
2                17.277920   1  99
3                17.179756   1  99
4                17.811403   1  99
5                15.213779   1  99
6                16.439806   1  99
7                17.225773   1  99
8                15.340520   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170828   1  99
12               15.887622   1  99
13               16.583304   1  99
14               16.757377   1  99
15               16.165037   1  99
16               17.663250   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

In [20]:
# Finding minimum mean
ords_prods_merge.groupby('department_id')['order_number'].mean().min()

15.213778838711356

In [22]:
# Finding maximum mean
ords_prods_merge.groupby('department_id')['order_number'].mean().max()

22.90237905850025

OBSERVATIONS:
1. Minimum mean value of the entire dataframe(15.2) is more than the subset(14.3).
2. Maximum mean value of the entire dataframe(22.9) is less than the subset(25.5).
3. Each department_id has seen some level of change in the final output. Most of the department_ids' values changed in decimals, however there are few department_ids like 10(+1.5),12(+1.5), and 22(-2.6), saw noticiable changes when the command in used on the entire dataframe.
4. The minimum value remained same in both the case. The minimum order_number is 1.
5. The maximum value also remained same in both the cases except for three department_id. Those ids are 2, 8 and 21. When aggregated the entire dataset the max order_number for every department_id is 99. Howeever, in subset the max order_number for department_id 2, 8 and 21 were 96, 98 and 98 respectively.

# 4. Aggregating Data with Transform()

In [26]:
# Creating Flag of "Loyal Customers"
ords_prods_merge['max_order']=ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/38/3t8_0dvn1h950p6h5yqb56rm0000gn/T/ipykernel_1912/2395488788.py:2: FutureWarning: The provided callable <function max at 0x10dfb5da0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order']=ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [27]:
ords_prods_merge.head(100)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
..         ...                         ...       ...            ...     ...   
95           1  Chocolate Sandwich Cookies        61             19     5.8   
96           1  Chocolate Sandwich Cookies        61             19     5.8   
97           1  Chocolate Sandwich Cookies        61             19     5.8   
98           1  Chocolate Sandwich Cookies        61             19     5.8   
99           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    3139998      138            28          6                 11   
1    1977647      138            30          6                 17   
2     389851      709             2          0                 21   
3     652770      764             1          3                 13   
4    1813452      764             3          4                 17   
..       ...      ...           ...        ...                ...   
95    602103    10831             8          3                 11   
96     49629    11119            18          1                 14   
97    317888    11186            13          5                 16   
98    682486    11243            16          3                 13   
99   2861386    11448            10          2                 15   

    days_since_prior_order  add_to_cart_order  reordered _merge  busiest_days  \
0                      3.0                  5          0   both  Regular Days   
1                     20.0                  1          1   both  Regular Days   
2                      6.0                 20          0   both  Busiest Days   
3                      NaN                 10          0   both  Slowest Days   
4                      9.0                 11          1   both  Slowest Days   
..                     ...                ...        ...    ...           ...   
95                    23.0                  5          0   both  Slowest Days   
96                    30.0                  1          0   both  Busiest Days   
97                     2.0                  8          0   both  Regular Days   
98                     0.0                  2          0   both  Slowest Days   
99                     3.0                  6          0   both  Regular Days   

   busiest_period_of_day  max_order  
0            Most orders         32  
1         Average Orders         32  
2         Average Orders          5  
3            Most orders          3  
4         Average Orders          3  
..                   ...        ...  
95           Most orders         10  
96           Most orders         23  
97           Most orders         26  
98           Most orders         43  
99           Most orders         21  

[100 rows x 17 columns]

In [28]:
pd.options.display.max_rows=None

In [32]:
ords_prods_merge.head(100)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   
20           1  Chocolate Sandwich Cookies        61             19     5.8   
21           1  Chocolate Sandwich Cookies        61             19     5.8   
22           1  Chocolate Sandwich Cookies        61             19     5.8   
23           1  Chocolate Sandwich Cookies        61             19     5.8   
24           1  Chocolate Sandwich Cookies        61             19     5.8   
25           1  Chocolate Sandwich Cookies        61             19     5.8   
26           1  Chocolate Sandwich Cookies        61             19     5.8   
27           1  Chocolate Sandwich Cookies        61             19     5.8   
28           1  Chocolate Sandwich Cookies        61             19     5.8   
29           1  Chocolate Sandwich Cookies        61             19     5.8   
30           1  Chocolate Sandwich Cookies        61             19     5.8   
31           1  Chocolate Sandwich Cookies        61             19     5.8   
32           1  Chocolate Sandwich Cookies        61             19     5.8   
33           1  Chocolate Sandwich Cookies        61             19     5.8   
34           1  Chocolate Sandwich Cookies        61             19     5.8   
35           1  Chocolate Sandwich Cookies        61             19     5.8   
36           1  Chocolate Sandwich Cookies        61             19     5.8   
37           1  Chocolate Sandwich Cookies        61             19     5.8   
38           1  Chocolate Sandwich Cookies        61             19     5.8   
39           1  Chocolate Sandwich Cookies        61             19     5.8   
40           1  Chocolate Sandwich Cookies        61             19     5.8   
41           1  Chocolate Sandwich Cookies        61             19     5.8   
42           1  Chocolate Sandwich Cookies        61             19     5.8   
43           1  Chocolate Sandwich Cookies        61             19     5.8   
44           1  Chocolate Sandwich Cookies        61             19     5.8   
45           1  Chocolate Sandwich Cookies        61             19     5.8   
46           1  Chocolate Sandwich Cookies        61             19     5.8   
47           1  Chocolate Sandwich Cookies        61             19     5.8   
48           1  Chocolate Sandwich Cookies        61             19     5.8   
49           1  Chocolate Sandwich Cookies        61  

# 5. Deriving Columns with loc()

In [35]:
# Creating Loyalty Flag 
ords_prods_merge.loc[ords_prods_merge['max_order']>40,'loyalty_flag']='Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order']<=40)&(ords_prods_merge['max_order']>10),'loyalty_flag']='Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order']<=10,'loyalty_flag']='New customer'

In [36]:
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: count, dtype: int64

In [39]:
# Checking number of column after creating loyalty flag
ords_prods_merge.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'order_dow', 'order_hour_of_day',
       'days_since_prior_order', 'add_to_cart_order', 'reordered', '_merge',
       'busiest_days', 'busiest_period_of_day', 'max_order', 'loyalty_flag'],
      dtype='object')

In [41]:
ords_prods_merge[['user_id','loyalty_flag','order_number']].head(60)

user_id      loyalty_flag  order_number
0       138  Regular customer            28
1       138  Regular customer            30
2       709      New customer             2
3       764      New customer             1
4       764      New customer             3
5       777  Regular customer            16
6       825      New customer             3
7       910  Regular customer            12
8      1052  Regular customer            10
9      1052  Regular customer            15
10     1379      New customer             1
11     1480    Loyal customer            36
12     1480    Loyal customer            38
13     1480    Loyal customer            39
14     1494  Regular customer             6
15     1494  Regular customer             7
16     1494  Regular customer             9
17     1540  Regular customer             4
18     1540  Regular customer             5
19     1540  Regular customer             6
20     1540  Regular customer             7
21     1540  Regular customer             8
22     1540  Regular customer             9
23     1540  Regular customer            14
24     1540  Regular customer            15
25     1540  Regular customer            18
26     1540  Regular customer            19
27     1540  Regular customer            20
28     1540  Regular customer            21
29     1540  Regular customer            22
30     1540  Regular customer            24
31     1540  Regular customer            25
32     1540  Regular customer            27
33     1540  Regular customer            28
34     1598      New customer             1
35     1598      New customer             2
36     1598      New customer             3
37     1647      New customer             1
38     1647      New customer             4
39     1647      New customer             6
40     1647      New customer             8
41     1647      New customer             9
42     2833      New customer             1
43     2850  Regular customer             7
44     2857  Regular customer            24
45     3010  Regular customer             5
46     3029      New customer             1
47     3395      New customer             4
48     3904      New customer             1
49     4006    Loyal customer             2
50     4006    Loyal customer             3
51     4006    Loyal customer             5
52     4006    Loyal customer             9
53     4006    Loyal customer            12
54     4122  Regular customer            11
55     4141      New customer             2
56     4343      New customer             1
57     4553      New customer             3
58     4845  Regular customer             6
59     6866      New customer             1

# 6. Question 5.

In [44]:
ords_prods_merge[['loyalty_flag','prices']].head(100)

loyalty_flag  prices
0   Regular customer     5.8
1   Regular customer     5.8
2       New customer     5.8
3       New customer     5.8
4       New customer     5.8
5   Regular customer     5.8
6       New customer     5.8
7   Regular customer     5.8
8   Regular customer     5.8
9   Regular customer     5.8
10      New customer     5.8
11    Loyal customer     5.8
12    Loyal customer     5.8
13    Loyal customer     5.8
14  Regular customer     5.8
15  Regular customer     5.8
16  Regular customer     5.8
17  Regular customer     5.8
18  Regular customer     5.8
19  Regular customer     5.8
20  Regular customer     5.8
21  Regular customer     5.8
22  Regular customer     5.8
23  Regular customer     5.8
24  Regular customer     5.8
25  Regular customer     5.8
26  Regular customer     5.8
27  Regular customer     5.8
28  Regular customer     5.8
29  Regular customer     5.8
30  Regular customer     5.8
31  Regular customer     5.8
32  Regular customer     5.8
33  Regular customer     5.8
34      New customer     5.8
35      New customer     5.8
36      New customer     5.8
37      New customer     5.8
38      New customer     5.8
39      New customer     5.8
40      New customer     5.8
41      New customer     5.8
42      New customer     5.8
43  Regular customer     5.8
44  Regular customer     5.8
45  Regular customer     5.8
46      New customer     5.8
47      New customer     5.8
48      New customer     5.8
49    Loyal customer     5.8
50    Loyal customer     5.8
51    Loyal customer     5.8
52    Loyal customer     5.8
53    Loyal customer     5.8
54  Regular customer     5.8
55      New customer     5.8
56      New customer     5.8
57      New customer     5.8
58  Regular customer     5.8
59      New customer     5.8
60      New customer     5.8
61  Regular customer     5.8
62  Regular customer     5.8
63  Regular customer     5.8
64  Regular customer     5.8
65  Regular customer     5.8
66  Regular customer     5.8
67  Regular customer     5.8
68  Regular customer     5.8
69    Loyal customer     5.8
70  Regular customer     5.8
71      New customer     5.8
72      New customer     5.8
73      New customer     5.8
74      New customer     5.8
75  Regular customer     5.8
76  Regular customer     5.8
77  Regular customer     5.8
78  Regular customer     5.8
79  Regular customer     5.8
80  Regular customer     5.8
81  Regular customer     5.8
82  Regular customer     5.8
83  Regular customer     5.8
84  Regular customer     5.8
85  Regular customer     5.8
86  Regular customer     5.8
87  Regular customer     5.8
88      New customer     5.8
89  Regular customer     5.8
90  Regular customer     5.8
91  Regular customer     5.8
92  Regular customer     5.8
93  Regular customer     5.8
94  Regular customer     5.8
95      New customer     5.8
96  Regular customer     5.8
97  Regular customer     5.8
98    Loyal customer     5.8
99  Regular customer     5.8

In [46]:
#Performing Multiple Aggregations
ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean','min','max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.388747  1.0  99999.0
New customer      13.294370  1.0  99999.0
Regular customer  12.496203  1.0  99999.0

According to the analysis, the loyal customers on an average spend less money on their shopping in comparison to the new and regular customers. The average price value of loyal customers is 10.3, which is followed by regular customers average price value 12.4 and the new customers on an average tend to spend 13.2 on their shopping. This also signifies that frequent buyers like to target low price goods as they buy regularly, whereas new customers like to shop high-values goods as they might be targeting on the duration of the goods and don't want to just concentrate on low-price goods. 

The minimum and maximum values in all categories are same. The max price values look very high according to this analysis. May be this data contains an outlier. 

# 7.Question 6

In [78]:
# Finding basic info about Price column
ords_prods_merge['prices'].describe()

count    3.243421e+07
mean     1.198114e+01
std      4.954311e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

In [80]:
# Creating "mean_product_price" column in the dataset
ords_prods_merge['mean_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/38/3t8_0dvn1h950p6h5yqb56rm0000gn/T/ipykernel_1912/2870687859.py:2: FutureWarning: The provided callable <function mean at 0x10dfb67a0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['mean_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [82]:
# Printing first 15 rows of the newly formed column in the dataset
ords_prods_merge[['user_id', 'mean_product_price']].head(15)

user_id  mean_product_price
0       138            6.935811
1       138            6.935811
2       709            7.930208
3       764            4.972414
4       764            4.972414
5       777            6.935398
6       825            5.957576
7       910            6.680000
8      1052            7.162500
9      1052            7.162500
10     1379            5.813636
11     1480            5.139394
12     1480            5.139394
13     1480            5.139394
14     1494            6.803797

In [84]:
# Creating "Spending_flag" and their value_counts
ords_prods_merge.loc[ords_prods_merge['mean_product_price'] >= 10, 'spending_flag'] = 'High spender'
ords_prods_merge.loc[ords_prods_merge['mean_product_price'] < 10, 'spending_flag'] = 'Low spender'
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31798751
High spender      635461
Name: count, dtype: int64

# 8.Question 7

In [87]:
# Creating "median_days_between_orders" columns 
ords_prods_merge['median_days_between_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/38/3t8_0dvn1h950p6h5yqb56rm0000gn/T/ipykernel_1912/889895459.py:2: FutureWarning: The provided callable <function median at 0x114eadee0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_days_between_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [89]:
# Printing first 15 rows of the newly formed column
ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders']].head(15)

user_id  days_since_prior_order  median_days_between_orders
0       138                     3.0                         8.0
1       138                    20.0                         8.0
2       709                     6.0                         8.0
3       764                     NaN                         9.0
4       764                     9.0                         9.0
5       777                    26.0                        11.0
6       825                    30.0                        20.0
7       910                    30.0                         6.0
8      1052                    19.0                        10.0
9      1052                    15.0                        10.0
10     1379                     NaN                        20.0
11     1480                     7.0                         7.0
12     1480                     7.0                         7.0
13     1480                    12.0                         7.0
14     1494                     2.0                         7.0

In [91]:
# Creating 'order_frequency' flag
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] > 20, 'order_frequency'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['median_days_between_orders'] <= 20) & (ords_prods_merge['median_days_between_orders'] > 10), 'order_frequency'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] <= 10, 'order_frequency'] = 'Frequent customer'

In [93]:
# Calculating value_counts of order_frequency column
ords_prods_merge['order_frequency'].value_counts(dropna = False)

order_frequency
Frequent customer        21577409
Regular customer          7217134
Non-frequent customer     3639669
Name: count, dtype: int64

In [95]:
# Printing first 15 rows of the relevant columns
ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders', 'order_frequency']].head(15)

user_id  days_since_prior_order  median_days_between_orders  \
0       138                     3.0                         8.0   
1       138                    20.0                         8.0   
2       709                     6.0                         8.0   
3       764                     NaN                         9.0   
4       764                     9.0                         9.0   
5       777                    26.0                        11.0   
6       825                    30.0                        20.0   
7       910                    30.0                         6.0   
8      1052                    19.0                        10.0   
9      1052                    15.0                        10.0   
10     1379                     NaN                        20.0   
11     1480                     7.0                         7.0   
12     1480                     7.0                         7.0   
13     1480                    12.0                         7.0   
14     1494                     2.0                         7.0   

      order_frequency  
0   Frequent customer  
1   Frequent customer  
2   Frequent customer  
3   Frequent customer  
4   Frequent customer  
5    Regular customer  
6    Regular customer  
7   Frequent customer  
8   Frequent customer  
9   Frequent customer  
10   Regular customer  
11  Frequent customer  
12  Frequent customer  
13  Frequent customer  
14  Frequent customer

# 8. Question 9

In [76]:
# Exporting as pickle file
ords_prods_merge.to_pickle(os.path.join(path,'02 DATA','PREPARED DATA','Orders_Products_Merged_Grouped_AggregatedDATA.pkl'))